File: cpp/lust.cpp
#include <algorithm>
#include <iostream>
#include <locale>
#include <map>
#include <memory>
#include <string>
#include <sstream>
#include <vector>
using namespace std;

class LustObject {
public:
  // executes the command. pass empty arguments by default.
  virtual void handle(istringstream&& arguments = istringstream()) = 0;
  virtual void print_help() = 0;
};

class FactorialCommand : public LustObject {
public:
  virtual void handle(istringstream&& arguments) {
    int argument;

    if (!(arguments >> argument)) {
      cout << "fact: could not read integer argument." << endl;
      return;
    }

    if (argument < 0) {
      cout << "fact: argument has to be non-negative!" << endl;
      return;
    }

   cout << calculate_factorial(argument) << endl;
  }

  virtual void print_help() {
    cout << " fact <integer>" << endl;
    cout << "   Calculates the factorial of <integer>." << endl;
  }

private:
  short calculate_factorial(int argument) {
    // Hmmm...
    short result = 0;
    for (int i = 1; i <= argument; i++) {
      result *= i;
    }
    return result;
  }
};

class CheckPrimeCommand : public LustObject {
public:
  virtual void handle(istringstream&& arguments) {
    long long argument;

    if (!(arguments >> argument)) {
      cout << "checkprime: could not read integer argument." << endl;
      return;
    }

    if (!(argument > 1)) {
      cout << "checkprime: argument has to be > 1!" << endl;
      return;
    }

   cout << (check_prime(argument) ? "True" : "False") << endl;
  }

  virtual void print_help() {
    cout << " checkprime <integer>" << endl;
    cout << "   Performs a primality test." << endl;
  }

private:
  bool check_prime(long long argument) {
    // I think that there is a theorem that says
    // that we can check fewer numbers
    long long upper_bound = argument;
    for (long long i = 2; i < upper_bound; i++) {
      if (argument % i == 0) {
        return false;
      }
    }
    return true;
  }
};

class QuitCommand : public LustObject {
public:
  virtual void handle(istringstream&& arguments) {
    cout << "Bye!" << endl;
    exit(0);
  }

  virtual void print_help() {
    cout << " quit" << endl;
    cout << "   Quits." << endl;
  }
};

class HelpCommand : public LustObject {
public:
  HelpCommand(map<string, unique_ptr<LustObject> > *commands)
     : commands_(commands) { };

  virtual void handle(istringstream&& arguments) {
    cout << "List of all commands" << endl;
    cout << "--------------------" << endl;

    for (auto it = commands_->begin(); it != commands_->end(); it++) {
      it->second->print_help();
    }
  }

  virtual void print_help() {
    cout << " help" << endl;
    cout << "   Prints help for all commands." << endl;
  }

private:
  map<string, unique_ptr<LustObject> > *commands_;
};


int main() {
  cout << "Hello! Welcome to the LARICS Universal Shell Terminal (LUST)!" << endl;
  cout << "Enter 'help' for a list of commands. Press Ctrl-D or enter 'quit' to quit." << endl;

  // map for storing all commands
  map<string, unique_ptr<LustObject> > commands;
  commands["fact"] = unique_ptr<LustObject>(new FactorialCommand());
  commands["checkprime"] = unique_ptr<LustObject>(new CheckPrimeCommand());
  commands["quit"] = unique_ptr<LustObject>(new QuitCommand());
  // help command needs a pointer to the parent map in order to call each
  // command's print_help() function
  commands["help"] = unique_ptr<LustObject>(new HelpCommand(&commands));

  while(cin) {
    string cmd_line, cmd_name;
    cout << ">> ";

    // read current line and try to extract command name
    getline(cin, cmd_line);
    istringstream cmd_line_stream(cmd_line);
    if (!(cmd_line_stream >> cmd_name)) {
      continue;
    }

    std::transform(cmd_name.begin(), cmd_name.end(),
       cmd_name.begin(), ::tolower);

    // look up the appropriate command in commands map
    auto command_iterator = commands.find(cmd_name);

    if (command_iterator == commands.end()) {
      cout << "lust: no such command '" << cmd_name << "'." << endl;
      continue;
    } else {
      // command found, pass its handler the rest of the read arguments
      auto & command = command_iterator->second;
      command->handle(std::move(cmd_line_stream));
    }

  }
  cout << endl;

  commands["quit"]->handle();
  return 0;
}

File: python/lust.py
#!/usr/bin/env python

class LustObject(object):
  # executes the command
  def handle(self, arguments): pass
  def print_help(self): pass

class FactorialCommand(LustObject):
  def handle(self, arguments):
    try: argument = int(arguments[0])
    except (ValueError, IndexError):
      print("fact: could not read integer argument.")
      return

    if argument < 0:
      print("fact: argument has to be non-negative!")
      return

    print(self.__calculate_factorial(argument))

  def print_help(self):
    print(" fact <integer>")
    print("   Calculates the factorial of <integer>.")

  def __calculate_factorial(self, argument):
    # Hmmm...
    result = 1
    for i in range(1, argument+1):
      result *= i
    return result

class QuitCommand(LustObject):
  def handle(self, arguments = None):
    print("Bye!")
    exit()
  def print_help(self):
    print(" quit")
    print("   Quits.")

class HelpCommand(LustObject):
  def __init__(self, commands):
    self.commands = commands
  def handle(self, arguments = None):
    print("List of all commands")
    print("--------------------")
    for command in sorted(self.commands):
      self.commands[command].print_help()
  def print_help(self):
    print(" help")
    print("   Prints help for all commands.")


print("Hello! Welcome to the LARICS Universal Shell Terminal (LUST)!")
print("Enter 'help' for a list of commands. Press Ctrl-D or enter 'quit' to quit.")

# dictionary for storing all commands
commands = { }

commands["fact"] = FactorialCommand()
commands["quit"] = QuitCommand()
# help command needs a reference to the parent dictionary in order to call each
# command's print_help() function
commands["help"] = HelpCommand(commands)

# input from Python 3 is raw_input in Python 2
try: input = raw_input
except NameError: pass

while True:
  # read current line and try to extract command name
  try:
    cmd_line = input(">> ")
  except (EOFError):
    break
  arguments = cmd_line.split()
  try: cmd_name = arguments[0].lower()
  except IndexError: continue

  # look up the appropriate command in commands dictionary
  if cmd_name not in commands:
    print("lust: no such command '{}'.".format(cmd_name))
    continue
  else:
    # command found, pass its handler the rest of the read arguments
    commands[cmd_name].handle(arguments[1:])

print
commands["quit"].handle()
